Add policy execution mode to metadata

flavio opened the pull request:
Enrich the policy metadata with a new field that explains which execution mode (kubewarden, opa, gatekeeper) must be used to evaluate the policy.
    
By default, when nothing is specified, we fall back to the kubewarden execution mode.

Note well, I reused the `PolicyExecutionMode` enum to identify the type of execution mode to be used. That forced me to implement a custom serializer and deserializer for the enum. That introduces some extra code (especially the deserialization bit), but that's how serde wants things to be done.

ereslibre commented on src/policy_evaluator.rs:
```diff
@@ -16,12 +16,76 @@ use crate::cluster_context::ClusterContext;
+        S: serde::ser::Serializer,
+    {
+        let mode = match self {
+            PolicyExecutionMode::KubewardenWapc => "kubewarden",
```
We could use `"kubewarden-wapc"` and be future proof.

flavio replied:
Done

ereslibre commented on src/policy_evaluator.rs:
```diff
@@ -16,12 +16,76 @@ use crate::cluster_context::ClusterContext;
+
+struct StringVisitor;
+
+impl<'de> serde::de::Visitor<'de> for StringVisitor {
+    type Value = String;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "a string")
+    }
+
+    fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        Ok(s.to_owned())
+    }
+}
```
Do we need this pattern?

flavio replied:
You need it if you want to implement a custom Derive. However, it turns out we don't need it, more below

ereslibre commented on src/policy_evaluator.rs:
```diff
@@ -16,12 +16,76 @@ use crate::cluster_context::ClusterContext;
+    }
+}
+
+impl<'de> serde::de::Deserialize<'de> for PolicyExecutionMode {
+    fn deserialize<D>(deserializer: D) -> Result<PolicyExecutionMode, D::Error>
+    where
+        D: serde::de::Deserializer<'de>,
+    {
+        let mode = deserializer.deserialize_str(StringVisitor)?;
+        match mode.as_str() {
+            "kubewarden" => Ok(PolicyExecutionMode::KubewardenWapc),
+            "opa" => Ok(PolicyExecutionMode::Opa),
+            "gatekeeper" => Ok(PolicyExecutionMode::OpaGatekeeper),
+            _ => Err(serde::de::Error::custom(format!(
+                "{} is not a valid execution mode",
+                mode
+            ))),
+        }
+    }
+}
```
Maybe I'm missing something, but do we need to implement this? Cannot we derive it?

ereslibre commented on src/policy_evaluator.rs:
```diff
@@ -16,12 +16,76 @@ use crate::cluster_context::ClusterContext;
+    }
+}
+
+impl fmt::Debug for PolicyExecutionMode {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mode = match self {
+            PolicyExecutionMode::KubewardenWapc => "kubewarden",
+            PolicyExecutionMode::Opa => "opa",
+            PolicyExecutionMode::OpaGatekeeper => "gatekeeper",
+        };
+
+        f.debug_struct("PolicyExecutionMode")
+            .field("mode", &mode)
+            .finish()
+    }
+}
```
Same here, cannot we derive this one?

ereslibre reviewed (commented):
Thank you, some comments.

flavio commented:
@ereslibre I've updated the code to use derive the serialize and deserialize functions. I think everything looks fine right now, because our enums are unit-like (they are not tuples). I think the serialization output would change once we change that.

Because of that, I prefer to leave the previous commit into the history, plus leave the unit tests I added.

ereslibre reviewed (approved):
Thank you 🙏
